GCE and AWS provisioners, dynamic provisioning: admins can configure zone(s) where PVs shall be created

pospispa opened the pull request:
Zone configuration capabilities for GCE and AWS dynamic provisioners are extended.
Admins can configure in a storage class a comma separated list of allowed zone(s).

Partly fixes Trello cards:
- [GCE provisioner, parse pvc.Selector](https://trello.com/c/CyemTzsK/259-finish-gce-provisioner-parse-pvc-selector-dynamic-provision)
- [AWS provisioner, parse pvc.Selector](https://trello.com/c/2XjouSWw/260-finish-aws-provisioner-parse-pvc-selector-dynamic-provision)

```release-note
GCE and AWS dynamic provisioners extension: admins can configure zone(s) in which a persistent volume shall be created.
```

cc: @jsafrane

Commit:
Added func ZonesToSet

An admin shall be able to configure a comma separated list of zones for a StorageClass.

That's why the func ZonesToSet (string) (set.String, error) is added. The func ZonesToSet converts a string containing a comma separated list of zones to a set. In case the list contains an empty zone an error is returned.

Commit:
Admin Can Specify in Which GCE Availability Zone(s) a PV Shall Be Created

An admin wants to specify in which GCE availability zone(s) users may create persistent volumes using dynamic provisioning.

That's why the admin can now configure in StorageClass object a comma separated list of zones. Dynamically created PVs for PVCs that use the StorageClass are created in one of the configured zones.

Commit:
Admin Can Specify in Which AWS Availability Zone(s) a PV Shall Be Created

An admin wants to specify in which AWS availability zone(s) users may create persistent volumes using dynamic provisioning.

That's why the admin can now configure in StorageClass object a comma separated list of zones. Dynamically created PVs for PVCs that use the StorageClass are created in one of the configured zones.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38505)
<!-- Reviewable:end -->

jsafrane commented:
Adding @kubernetes/sig-storage.

This PR changes semantics of GCE and AWS `StorageClass.Parameters['zone']` from "the single zone where every volume of this class should be provisioned" to "comma separated list of zones, one will be chosen randomly during provisioning of each PV".

@pospispa, can you please update examples/experimental/persistent-volume-provisioning/README.md ?

jsafrane commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -103,16 +103,19 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
+	} else if zones, err = volume.Zones2Set(zone); err != nil {
+		return "", 0, nil, err
 	}
+	zone = volume.ChooseZoneForVolume(zones, c.options.PVC.Name)
```
I don't like reusing `zone` once for comma-separated list of zones (above) and once for the chosen zone (here).

thockin replied:
I sort of agree, on the other hand 2 flags need mutual-exclusion checking and error handling...

jsafrane commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1590,6 +1590,10 @@ func (c *Cloud) CreateDisk(volumeOptions *VolumeOptions) (KubernetesVolumeID, er
 	createAZ := volumeOptions.AvailabilityZone
```
I would be better to rename `AvailabilityZone` to `AvailabilityZones` to make it clear what's expected here.

jsafrane commented:
Alternatively, we could add new parameter "zones" with comma-separated list and leave "zone" for backward compatibility (i.e. until v2.0)

jsafrane commented:
@k8s-bot ok to test

thockin commented:
@jingxu97 can you take this for me - it requires a "taste" decision, I think.

@matchstick FYI

jingxu97 commented:
Agreed with @jsafrane for adding notation for "zones" 
Also, according to the design doc https://github.com/pmorie/kubernetes/blob/7aa61dd0ff3908784acb4fa300713f02e62119af/docs/proposals/volume-provisioning.md, claim.Spec.Selector and claim.Spec.Class  are mutually exclusive. How can a user specify the zone in selection to match storageclass's zones?

jsafrane commented:
@jingxu97, claim.Spec.Selector and claim.Spec.Class were mutually exclusive to simplify the implementation, now we're "fixing" it by allowing certain labels in selectors together with class so a user can configure zone where a volume should be provisioned. volume-provisioning proposal has been updated with this information a while ago, see current version at https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/volume-provisioning.md#design.

I'm removing the note about mutual exclusivity in #38964

pospispa commented:
@k8s-bot verify test this

pospispa commented:
@k8s-bot bazel test this

jsafrane commented:
/lgtm

jsafrane commented:
@saad-ali @jingxu97 @pmorie, PTAL. It looks good to me and it will be even improved in #37497

jsafrane commented on test/test_owners.csv:
```diff
@@ -696,7 +697,6 @@ k8s.io/kubernetes/pkg/kubelet/kuberuntime,yifan-gu,1
 k8s.io/kubernetes/pkg/kubelet/lifecycle,yujuhong,1
 k8s.io/kubernetes/pkg/kubelet/network,freehan,0
 k8s.io/kubernetes/pkg/kubelet/network/cni,freehan,0
-k8s.io/kubernetes/pkg/kubelet/network/exec,freehan,0
```
This looks suspicious, this PR does not remove any tests. Can you please remove this line?

pospispa commented:
@jsafrane  @saad-ali @jingxu97 @pmorie @justinsb, PTAL. I believe all comments were addressed and the PR is ready to be merged.

saad-ali commented:
Sorry for the delay here, we'll take a look.

jsafrane commented:
I'm adding milestone 1.7, IMO it's too late for 1.6 and we need design in https://github.com/kubernetes/community/pull/247 merged first.

jsafrane commented:
oops, sorry, wrong PR. This one does not add selectors.

pospispa commented:
@k8s-bot node e2e test this

pospispa commented:
@k8s-bot cvm gke e2e test this

pospispa commented:
@k8s-bot gci gke e2e test this

msau42 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -287,11 +287,14 @@ const (
+	Tags              map[string]string
+	PVCName           string
+	VolumeType        string
+	ZonePresent       bool
```
Do you need these two booleans?  Can you just infer from if the string is empty?

pospispa replied:
IMHO, I can approach it using one of the below ways:

A, riskier: infer from the empty string `AvailabilityZone` or `AvailabilityZones` that the optional parameter `zone` or `zones` wasn't present. As the Persistent Volume Claim parameters are provided as `map[string]string` the values `"zone": ""` or `"zones": ""` are perfectly valid values in Go lang even though K8s yaml file parsing shall never provide such values. So in case I relay on the K8s yaml file parsing I can use the `AvailabilityZone` or `AvailabilityZones` empty string values as an indication that the `zone` or `zones` parameter were not specified in a Persistent Volume Claim. However, if my code is "fed" with unexpected input the incorrect input value `"zone": ""` or `"zones": ""` will remain undetected.

B, defensive: use a dedicated booleans to indicate whether the `zone` or `zones` optional parameters were present or not.

I chose the defensive approach.

msau42 replied:
So if I understand correctly, zone or zones == "" is invalid, and should return an error, while zone && zones == nil is valid.  Can you add unit tests for all these permutations of zone/zones == nil, "", and value?

pospispa replied:
The check for `zones == ""` was already present in the `func ZonesToSet`.

I added `func ValidateZone` to check for `zone == ""`.

Commit:
Added func ValidateZone

The zone parameter provided in a Storage Class may erroneously be an empty string or contain only spaces and tab characters. Such situation shall be detected and reported as an error.

That's why the func ValidateZone was added.

jingxu97 commented:
@saad-ali can we put this into 1.6?

enisoc commented:
I'm cleaning out the milestone since we're coming up on the cut. This doesn't seem to be a release-blocker.

pospispa commented:
@k8s-bot gce etcd3 e2e test this

pospispa commented:
@k8s-bot gci gke e2e test this

pospispa commented:
@k8s-bot cvm gke e2e test this

jsafrane commented:
/lgtm

jsafrane commented:
@thockin, please approve

pospispa commented:
@thockin please take a look.

pospispa commented:
@jingxu97 @jsafrane @saad-ali unfotunatelly, your LGTM was removed because of rebase. Please, will you re-approve it? In addition, there was an attempt to merge this PR into K8s 1.6, however, it was too late. @saad-ali it seems that the history may repeat for K8s 1.7.

@brendandburns @thockin @zmerlynn please, will you approve the small change in examples.

jsafrane commented:
/lgtm

pospispa commented:
/assign @thockin

jsafrane commented:
/lgtm

childsb commented:
@thockin can you /approve ? There's a tweak to /examples which takes super approval powers.

thockin commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -25,12 +25,13 @@ metadata:
 
 * `type`: `io1`, `gp2`, `sc1`, `st1`. See AWS docs for details. Default: `gp2`.
-* `zone`: AWS zone. If not specified, a random zone from those where Kubernetes cluster has a node is chosen.
+* `zone`: AWS zone. If neither zone nor zones is specified, a random zone in the same region as controller-manager will be chosen. Note: zone and zones parameters must not be used at the same time.
```
Can we re-spec the default as "a random zone chosen from the set of zones in which cluster VMs exist" ?

thockin commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -45,11 +46,12 @@ metadata:
 * `type`: `pd-standard` or `pd-ssd`. Default: `pd-ssd`
-* `zone`: GCE zone. If not specified, a random zone in the same region as controller-manager will be chosen.
+* `zone`: GCE zone. If neither zone nor zones is specified, a random zone in the same region as controller-manager will be chosen. Note: zone and zones parameters must not be used at the same time.
+* `zones`: a comma separated list of GCE zone(s). If neither zone nor zones is specified, a random zone in the same region as controller-manager will be chosen. Note: zone and zones parameters must not be used at the same time.
```
same comment

thockin commented on pkg/volume/util.go:
```diff
@@ -410,3 +410,27 @@ func JoinMountOptions(userOptions []string, systemOptions []string) []string {
+	zonesSlice := strings.Split(zonesString, ",")
+	zonesSet := make(sets.String)
+	for _, zone := range zonesSlice {
+		trimmedZone := strings.TrimSpace(zone)
```
If we defined it as a space separated string, we could just use strings.Fields() and maybe not need a helper function at all?

pospispa replied:
It is necessary to convert the input string with list of zones to a set because the [func ChooseZoneForVolume ](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util.go#L295) requires a set as one of the input parameters. As the `strings.Fields` converts a string to a slice a helper function will be needed anyway. However, the line `trimmedZone := strings.TrimSpace(zone)` could be deleted if `strings.Fields` is used.
IMHO, comma separated list feels better than a space separated list. That's why I prefer to keep it as it is.

jsafrane replied:
ack

thockin commented on pkg/volume/util.go:
```diff
@@ -410,3 +410,27 @@ func JoinMountOptions(userOptions []string, systemOptions []string) []string {
+// ValidateZone returns:
+// - an error in case zone is an empty string or contains only any combination of spaces and tab characters
+// - nil otherwise
+func ValidateZone(zone string) error {
```
Can we in-line this at call-sites?  It's sort of silly to out-of-line this.

pospispa replied:
IMHO, it's a matter of subjective preference.
I prefer to express my [intention](https://en.wikipedia.org/wiki/Intentional_programming) at call-sites rather than to in-line the implementation at call-sites even though the current implementation of the validation is trivial.
In addition, the function `ValidateZone` is not called at only 1 place.
That's why I prefer to keep it as it is.

thockin reviewed (commented):
I have a couple nits - I will approve, but remove LGTM.  @jsafrane and @pospispa and @msau can decide if I am wrong and re-LGTM.

thockin commented:
/approve

pospispa commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

pospispa commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

pospispa commented:
@k8s-bot cvm gke e2e test this

pospispa commented:
@k8s-bot gci gke e2e test this

jsafrane commented:
/lgtm
